src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -348,16 +417,365 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+        {
+            Contract.Assert(readContext != null);
+
+            ODataResource resource = new ODataResource
+            {
+                TypeName = edmPropertyType.FullName(),
+            };
+
+            resource.Properties = CreateKeyProperties(refLink.EntityReferenceLink.Url, readContext);
+
+            if (refLink.EntityReferenceLink.InstanceAnnotations != null)
+            {
+                foreach (ODataInstanceAnnotation instanceAnnotation in refLink.EntityReferenceLink.InstanceAnnotations)
+                {
+                    resource.InstanceAnnotations.Add(instanceAnnotation);
+                };
+            }
+
+            return new ODataResourceWrapper(resource);
+        }
+
+
+        /// <summary>
+        /// Do uri parsing to get the key values.
+        /// </summary>
+        /// <param name="id">The key Id.</param>
+        /// <param name="readContext">The reader context.</param>
+        /// <returns>The key properties.</returns>
+        private static IList<ODataProperty> CreateKeyProperties(Uri id, ODataDeserializerContext readContext)
+        {
+            Contract.Assert(id != null);
+            Contract.Assert(readContext != null);
+            IList<ODataProperty> properties = new List<ODataProperty>();
+            if (readContext.Request == null)
+            {
+                return properties;
+            }
+
+            ODataPath odataPath = GetODataPath(id.OriginalString, readContext);
+            if (odataPath != null)
+            {
+                KeySegment keySegment = odataPath.OfType<KeySegment>().LastOrDefault();
+
+                if (keySegment != null)
+                {
+                    foreach (KeyValuePair<string, object> key in keySegment.Keys)
+                    {
+                        properties.Add(new ODataProperty
+                        {
+                            Name = key.Key,
+                            Value = key.Value
+                        });
+                    }
+                }
+            }
+
+            return properties;
+        }
+
+        private static ODataPath GetODataPath(string id, ODataDeserializerContext readContext)
+        {
+            try
+            {
+                Routing.IODataPathHandler pathHandler = readContext.InternalRequest.PathHandler;
+                IWebApiRequestMessage internalRequest = readContext.InternalRequest;
+                IWebApiUrlHelper urlHelper = readContext.InternalUrlHelper;
+
+                string serviceRoot = urlHelper.CreateODataLink(
+                    internalRequest.Context.RouteName,
+                    internalRequest.PathHandler,
+                    new List<ODataPathSegment>());
+                ODataPath odataPath = pathHandler.Parse(serviceRoot, id, internalRequest.RequestContainer).Path;
+
+                return odataPath;
+            }
+            catch (Exception)
+            {
+                return null;
```

Comment:
Take into account the fix we had to apply to 8.x https://github.com/OData/AspNetCoreOData/pull/643/files
This catch block would swallow the exception

Reply:
Example of a swallowed exception I came across when testing. Typically in this scenario we should have performed a null check, instead of a `NullReferenceException` getting caught and ignored.
![image](https://user-images.githubusercontent.com/28291332/184598835-a89130b5-8f35-4a80-862b-b269885621a7.png)